service_provider.go

```diff
@@ -1534,6 +1531,13 @@ func (sp *ServiceProvider) ValidateLogoutResponseRedirect(queryParameterData str
 		return err
 	}
 
+	if query.Get("Signature") != "" && query.Get("SigAlg") != "" {
+		if err := sp.validateQuerySig(query); err != nil {
```

Comment:
I see here that the PR validates a signature coming via query elements.

However,  in line 1547 the code still calls `sp.validateSignature(doc.Root())`, which 
expects a signature element in the XML document, and wants to verify it.

So, this PR seems to expect a signature in two places, required in the XML document, and optional in the query.

With keycloak I have a signature in the query element, and nothing in the XML document.
I expect this to fail.

I am missing something ?

Reply:
Added a querySig flag for myself, recording when a proper query sig is seen.
Check this flag if xml sign errors out with not present to avoid passing on that error

IOW

```
	querySig := false
	if query.Get("Signature") != "" && query.Get("SigAlg") != "" {
		if err := sp.validateQuerySig(query); err != nil {
			retErr.PrivateErr = err
			return retErr
		}
		querySig = true
	}
```

and

```
	if err := sp.validateSignature(doc.Root()); err != nil {
		if err != errSignatureElementNotPresent || !querySig {
			retErr.PrivateErr = err
			return retErr
		}
 	}
```

Reply:
Tested the modification against Keycloak and works. As expected it verifies the detached sig, and then ignores the missing embedded sig.

Also extended further to generate a proper detached sig for logout requests. That solved issues with OKTA.

References:
* https://github.com/rancher/saml/commit/e1e76fd2b4b6fa44c3264311f093cc7bce14d045
* https://github.com/rancher/saml/commit/a10be19085301a0f36419b77a68cab400eaf9a7c